src/components/tl/mlx5/alltoall/alltoall_coll.c

```diff
@@ -0,0 +1,873 @@
+    ucc_tl_mlx5_alltoall_t *a2a     = tl_team->a2a;
+    int             is_asr = (a2a->node.sbgp->group_rank == a2a->node.asr_rank);
+    int             i, n_tasks = is_asr ? 4 : 2, curr_task = 0;
+    ucc_schedule_t         *schedule;
+    ucc_tl_mlx5_schedule_t *task;
+    size_t                  msg_size;
+    int                     block_size;
+    ucc_coll_task_t        *tasks[4];
+    ucc_status_t            status;
+
+    if (UCC_IS_INPLACE(coll_args->args)) {
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    msg_size = coll_args->args.src.info.count / UCC_TL_TEAM_SIZE(tl_team) *
+               ucc_dt_size(coll_args->args.src.info.datatype);
+
+    if (msg_size > a2a->max_msg_size) {
+        tl_debug(UCC_TL_TEAM_LIB(tl_team), "msg size too long");
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    status = ucc_tl_mlx5_get_schedule(tl_team, coll_args, &task);
+    if (ucc_unlikely(UCC_OK != status)) {
+        return status;
+    }
+    schedule = &task->super;
+
+    for (i = 0; i < n_tasks; i++) {
+        tasks[i] = &ucc_tl_mlx5_init_task(coll_args, team, schedule)->super;
+    }
+
+    task->alltoall.send_blocks_enqueued = 0;
+    task->alltoall.started              = 0;
+    task->alltoall.wait_wc              = 0;
+    task->alltoall.blocks_sent          = 0;
+    task->alltoall.blocks_completed     = 0;
+    task->alltoall.seq_num              = a2a->sequence_number;
+    task->alltoall.seq_index            = SEQ_INDEX(a2a->sequence_number);
+    task->alltoall.op       = &a2a->node.ops[task->alltoall.seq_index];
+    task->alltoall.msg_size = msg_size;
+
+    tl_debug(UCC_TL_TEAM_LIB(tl_team), "Seq num is %d", task->alltoall.seq_num);
+    a2a->sequence_number += 1;
+
+    block_size = a2a->requested_block_size ? a2a->requested_block_size
+                                           : get_block_size(task);
+
+    //todo following section correct assuming homogenous PPN across all nodes
+    task->alltoall.num_of_blocks_columns =
+        (a2a->node.sbgp->group_size % block_size)
+            ? ucc_div_round_up(a2a->node.sbgp->group_size, block_size)
+            : 0;
+    if (((a2a->net.sbgp->group_rank == 0) && is_asr) &&
+        (msg_size != a2a->previous_msg_size[task->alltoall.seq_index])) {
+        tl_info(UCC_TL_TEAM_LIB(tl_team), "Block size is %d msg_size is %zu",
+                block_size, msg_size);
+    }
+
+    task->alltoall.block_size = block_size;
+
+    // TODO remove for connectX-7 - this is mkey_entry->stride (count+skip) limitation - only 16 bits
+    if (task->alltoall
+            .num_of_blocks_columns) { // for other case we will never reach limit - we use bytes skip only for the "leftovers" mode, which means when
+                                      // num_of_blocks_columns != 0
+        size_t limit =
+            (1ULL
+             << 16); // TODO We need to query this from device (or device type) and not user hardcoded values
```

Comment:
why not to query then?

Reply:
I cannot find in the PRM how to query that, it is just that this field is 16bytes. So maybe we should add a query when CX8 is released if this limitation changes. In the meantime I would just leave like that, it is in my opinion very low priority -- if you disagree I need to ask Rami

Let me point out that for now this limitation only deals with the irregular case and is quite impossible to be reached: since msg_size<=128Bytes, it would require ppn >= 2^9 ...